README.rst
=========
autoemcee
=========

Runs MCMC automatically to convergence.

About
-----

Runs a family of Markov Chain Monte Carlo ensemble samplers (Affine-Invariant or Slice Sampler)
with gradually increasing number of samples until they converge.

Convergence is tested for each ensemble and for each parameter with the Geweke diagnostic.
Additionally, across ensembles, the Gelman-Rubin r-hat is used.

Can be run with MPI without modifying the code.

.. image:: https://img.shields.io/pypi/v/autoemcee.svg
        :target: https://pypi.python.org/pypi/autoemcee

.. image:: https://api.travis-ci.org/JohannesBuchner/autoemcee.svg?branch=master&status=started
        :target: https://travis-ci.org/github/JohannesBuchner/autoemcee

.. image:: https://img.shields.io/badge/docs-published-ok.svg
        :target: https://johannesbuchner.github.io/autoemcee/
        :alt: Documentation Status


This package is built on top of emcee, zeus, anviz and mpi4py.

You can help by testing autoemcee and reporting issues. Code contributions are welcome.
See the `Contributing page <https://johannesbuchner.github.io/autoemcee/contributing.html>`_.

Features
---------

* Pythonic. pip installable.
* Easy to program for: Sanity checks with meaningful errors
* Supports both emcee and zeus
* MPI support

Usage
^^^^^

Read the full documentation at:

https://johannesbuchner.github.io/autoemcee/


For parallelisation, use::

        mpiexec -np 4 python3 yourscript.py


Licence
^^^^^^^

GPLv3 (see LICENCE file). If you require another license, please contact me.

Icon made by `Vecteezy <https://www.flaticon.com/authors/smashicons>`_ from `Flaticon <https://www.flaticon.com/>`_ .


HISTORY.rst
==============
Release Notes
==============

0.1.0 (2020-03-07)
------------------

* First version



tests/test_run.py
import numpy as np
from autoemcee import ReactiveAffineInvariantSampler


def test_run():
    np.random.seed(1)

    paramnames = ['Hinz', 'Kunz']
    def loglike(z):
        assert len(z) == len(paramnames)
        a = -0.5 * (((z - 0.5) / 0.01)**2).sum() + -0.5 * ((z[0] - z[1])/0.01)**2
        loglike.ncalls += 1
        return a
    loglike.ncalls = 0

    def transform(x):
        assert len(x) == len(paramnames)
        return 10. * x - 5.

    sampler = ReactiveAffineInvariantSampler(paramnames, loglike, transform=transform, sampler='goodman-weare')
    r = sampler.run(max_improvement_loops=1)
    #sampler.laplace_approximate()
    
    ncalls = loglike.ncalls
    assert abs(sampler.ncall - ncalls) == 2, (sampler.ncall, ncalls)

def test_rosen():
    np.random.seed(1)

    paramnames = ['Hinz', 'Kunz', 'Fuchs', 'Gans', 'Hofer']
    def loglike(theta):
        assert len(theta) == len(paramnames)
        a = theta[:-1]
        b = theta[1:]
        return -2 * (100 * (b - a**2)**2 + (1 - a)**2).sum()

    def transform(u):
        assert len(u) == len(paramnames)
        return u * 20 - 10
    
    sampler = ReactiveAffineInvariantSampler(paramnames, loglike, transform=transform, sampler='slice')
    sampler.run(max_improvement_loops=1)
    

if __name__ == '__main__':
    test_run()
    test_rosen()


searchindex.js
Search.setIndex({docnames:["autoemcee","contributing","example","history","index","installation","issues","modules","readme"],envversion:{"sphinx.domains.c":1,"sphinx.domains.changeset":1,"sphinx.domains.citation":1,"sphinx.domains.cpp":1,"sphinx.domains.index":1,"sphinx.domains.javascript":1,"sphinx.domains.math":2,"sphinx.domains.python":1,"sphinx.domains.rst":1,"sphinx.domains.std":1,"sphinx.ext.viewcode":1,nbsphinx:2,sphinx:56},filenames:["autoemcee.rst","contributing.rst","example.ipynb","history.rst","index.rst","installation.rst","issues.rst","modules.rst","readme.rst"],objects:{"":{autoemcee:[0,0,0,"-"]},"autoemcee.ReactiveAffineInvariantSampler":{find_starting_walkers:[0,2,1,""],plot:[0,2,1,""],print_results:[0,2,1,""],run:[0,2,1,""]},autoemcee:{ReactiveAffineInvariantSampler:[0,1,1,""]}},objnames:{"0":["py","module","Python module"],"1":["py","class","Python class"],"2":["py","method","Python method"]},objtypes:{"0":"py:module","1":"py:class","2":"py:method"},terms:{"09it":2,"16it":2,"37it":2,"38it":2,"45it":2,"52it":2,"67it":2,"98it":2,"catch":0,"class":0,"float":0,"function":[0,1,2],"import":2,"int":0,"new":1,"public":5,"return":[0,2],"true":[0,2],"try":2,But:2,For:[0,4,8],The:[0,1,5],Then:1,Useful:0,Using:2,abl:2,about:1,abov:0,accept:2,across:[4,8],add:1,addition:[0,4,8],affin:[0,4,8],all:1,alreadi:2,alwai:[1,5],ani:1,anoth:[4,8],anviz:[4,8],anyth:1,api:4,appreci:1,arbitrari:0,arrai:0,articl:1,ask:4,assum:1,attempt:0,auto:[0,2],autocorrel:0,autoemce:[2,5,7],automat:[4,8],bandwidth:2,bar:0,base:0,bayesian:0,befor:1,best:1,beutler:0,bit:[1,2],blog:1,bool:0,both:[4,8],branch:1,bug:[0,6],bugfix:1,built:[4,8],bump2vers:1,burn:2,calcul:0,call:2,can:[1,2,4,5,8],cannot:2,carlo:[4,8],chain:[0,2,4,8],chang:1,check:[1,2,4,8],checkout:1,clone:[1,5],cmap:2,code:[4,8],color:2,com:[1,5],command:5,commit:1,comput:0,contact:[4,8],content:4,contour:2,contour_color:2,contribut:[4,6,8],converg:[0,2,4,8],copi:[1,5],corner:2,correl:0,could:1,creat:[1,2],credit:1,cube:0,curl:5,curvatur:2,debug:2,def:2,densiti:2,describ:2,descript:1,detail:1,develop:1,diagnost:[0,2,4,8],dict:2,did:2,dim:0,dimension:0,doc:1,docstr:1,document:8,don:5,done:1,download:5,draw:0,driven:1,each:[0,4,8],easi:[4,8],easier:1,effort:2,either:5,emce:[0,4,8],enhanc:1,enough:2,ensembl:[0,4,8],entri:1,error:[0,2,4,8],estim:2,evalu:0,even:[1,2],everi:1,evid:0,explain:1,explor:0,fail:2,fallback:2,fals:[0,2],famili:[4,8],few:2,file:[1,4,8],fill:2,find:2,find_starting_walk:0,fine_bins_2d:2,first:[0,3],fit:2,flake8:1,flaticon:[4,8],follow:2,fork:1,frequent:4,from:[0,2,4,8],fuch:2,full:[4,8],gan:2,gaussian:2,gave:2,gelman:[0,4,8],get_single_plott:2,get_subplot_plott:2,getdist:2,getlogg:2,gewek:[0,4,8],git:[1,5],github:[1,4,5,6,8],give:0,given:1,global:2,good:2,goodman:0,gplv3:[4,8],gradual:[4,8],greatli:1,guid:5,hat:[4,8],have:[0,1,5],help:[1,4,8],here:[1,2],highest:0,hint:6,hinz:2,histori:1,hofer:2,how:[1,2],http:[1,4,5,8],icon:[4,8],includ:1,increas:[0,4,8],independ:0,index:4,info:2,initi:2,initialis:0,instal:[1,4,8],invari:[0,4,8],issu:[1,4,6,8],iter:[0,2],johannesbuchn:[1,4,5,8],just:1,karamani:0,keep:1,kunz:2,kwarg:0,label:2,larg:2,larger:0,length:0,let:2,licens:[4,8],likelihood:[0,2],linspac:2,list:[0,1],littl:1,local:1,log:[0,2],loglik:[0,2],loglike_rosen:2,loglikelihood:2,logz:2,look:1,made:[4,8],maintain:1,major:1,make:1,mani:1,margin:0,markov:[4,8],master:5,matplotlib:2,max:0,max_improvement_loop:0,max_ncal:[0,2],maximum:0,mcmc:[0,2,4,8],mcsampl:2,meaning:[4,8],meet:1,meshgrid:2,method:5,might:1,min_autocorr_tim:0,minor:1,mkvirtualenv:1,model:0,modifi:[4,8],modul:[4,7],monomod:0,mont:[4,8],more:[1,2],most:5,mpi4pi:[4,8],mpi:[0,4,8],mpiexec:[4,8],multipl:0,n_eff:2,name:[0,1,2],narrow:1,ncall:2,none:0,note:4,now:1,num_chain:0,num_global_sampl:0,num_initial_step:0,num_plot_contour:2,num_step:0,num_test_sampl:0,num_walk:0,number:[0,4,8],numpi:2,object:0,offici:1,ojl:5,onc:5,open:[1,6],oper:1,optim:2,org:1,origin:1,other:1,packag:[4,8],page:[4,8],parallelis:[4,8],param_nam:0,paramet:[0,4,8],paramnam:2,part:1,pass:1,patch:1,perfectli:2,physic:0,pip:[1,4,5,8],pleas:[1,4,8],plot:[0,2],plot_2d:2,plt:2,point:[0,2],posit:0,possibl:1,post:1,posterior:0,prefer:5,print:2,print_result:0,prior:0,problem:0,process:[0,5],program:[4,8],progress:0,project:1,propos:1,pull_request:1,push:1,put:1,pypi:1,pyplot:2,pytest:1,python3:[4,8],python:[1,4,5,8],question:4,quit:2,random:0,rank:0,rate:2,reactiv:0,reactiveaffineinvariantsampl:[0,2],read:[4,8],readi:1,readm:1,receiv:0,recent:5,red:2,releas:4,rememb:1,remind:1,remov:2,repeat:0,repo:[1,5],report:[4,6,8],repositori:5,reproduc:1,requir:[0,4,8],result:[0,2],rhat:[0,2],root:2,rosenbrock:2,roughli:2,rst:1,rubin:[0,4,8],run:[0,1,2,4,5,8],sampl:[0,2,4,8],sampler:[0,2,4,8],samples_g:2,saniti:[4,8],scope:1,search:4,see:[0,2,4,6,8],select:0,send:1,set1:2,set:[0,1,2],setlevel:2,setup:[1,5],should:[0,1],show:0,show_titl:2,simpl:2,size:0,slice:[0,4,8],small:2,smooth_scale_2d:2,snowlin:1,sourc:0,start:2,step:[0,1,2],stop:0,str:0,strategi:0,subset:1,sum:2,summari:0,support:[1,4,8],sure:1,system:1,tag:1,take:0,tarbal:5,termin:5,test:[0,1,4,8],test_snowlin:1,than:0,thei:[1,4,8],them:1,theta:2,thi:[0,1,2,4,5,8],through:[1,5],time:[0,2],too:2,took:2,top:[4,8],tox:1,transform:[0,2],transform_rosen:2,travi:1,triangle_plot:2,troubleshoot:1,tutori:4,uncertainti:2,understand:2,unit:0,until:[0,4,8],updat:1,use:[0,1,4,8],used:[0,4,8],valid:2,vecteezi:[4,8],vector:0,veri:2,version:[1,3],virtualenv:1,virtualenvwrapp:1,volunt:1,wai:1,walker:0,want:1,warn:2,wear:0,web:1,websit:1,welcom:[1,8],well:2,when:1,whether:1,whoever:1,width_inch:2,without:[4,8],work:1,worst:2,would:1,xlim:2,ylim:2,you:[1,2,4,5,8],your:[1,5],yourscript:[4,8],zeu:[4,8]},titles:["autoemcee module","Contributing","Tutorial","Release Notes","Welcome to autoemcee\u2019s documentation!","Installation","Frequently Asked Questions","API","autoemcee"],titleterms:{about:[4,8],add:6,advanc:2,api:7,ask:6,autoemce:[0,4,8],bug:1,can:6,contribut:1,deploi:1,document:[1,4],featur:[1,4,8],feedback:1,fix:1,frequent:6,from:5,get:1,guidelin:1,here:6,how:6,implement:1,indic:4,instal:5,licenc:[4,8],modul:0,note:3,pull:1,question:6,quickstart:2,releas:[3,5],report:1,request:1,sourc:5,stabl:5,start:1,submit:1,tabl:4,tip:1,tutori:2,type:1,usag:[2,4,8],visualis:2,welcom:4,write:1}})

examples/sine_example.py
import argparse
import numpy as np
from numpy import pi, sin, log
import matplotlib.pyplot as plt
import logging, sys

def main(args):

    np.random.seed(2)
    Ndata = args.ndata
    jitter_true = 0.1
    phase_true = 2.0
    period_true = 180
    amplitude_true = args.contrast / Ndata * jitter_true
    paramnames = ['amplitude', 'jitter', 'phase', 'period']
    ndim = len(paramnames)
    
    x = np.linspace(0, 360, 1000)
    y = amplitude_true * sin(x / period_true * 2 * pi + phase_true)
    
    if True:
        plt.plot(x, y)
        x = np.random.uniform(0, 360, Ndata)
        y = np.random.normal(amplitude_true * sin(x / period_true * 2 * pi + phase_true), jitter_true)
        plt.errorbar(x, y, yerr=jitter_true, marker='x', ls=' ')
        plt.savefig('testsine.pdf', bbox_inches='tight')
        plt.close()
    
    
    def loglike(params):
        amplitude, jitter, phase, period = params
        predicty = amplitude * sin(x / period * 2 * pi + phase)
        logl = (-0.5 * log(2 * pi * jitter**2) - 0.5 * ((predicty - y) / jitter)**2).sum()
        return logl
    
    def transform(x):
        z = np.empty_like(x)
        z[0] = 10**(x[0] * 4 - 2)
        z[1] = 10**(x[1] * 1 - 1.5)
        z[2] = 2 * pi * x[2]
        z[3] = 10**(x[3] * 4 - 1)
        #z[:,4] = 2 * pi / x[:,3]
        return z

    loglike(transform(np.ones(ndim)*0.5))
    
    from autoemcee import ReactiveAffineInvariantSampler, create_logger
    
    create_logger('autoemcee', level=logging.DEBUG)
    sampler = ReactiveAffineInvariantSampler(paramnames, loglike, transform=transform)
    
    sampler.run(
        #num_global_samples=11,
        #num_chains=1,
        #num_walkers=10,
        #max_improvement_loops=40,
    )
    if sampler.mpi_rank != 0:
        return
    sampler.print_results()
    
    from getdist import MCSamples, plots

    samples_g = MCSamples(samples=sampler.results['samples'],
                           names=sampler.results['paramnames'],
                           label='Sine',
                           settings=dict(smooth_scale_2D=1))

    mcsamples = [samples_g]
    g = plots.get_subplot_plotter(width_inch=8)
    g.settings.num_plot_contours = 3
    g.triangle_plot(mcsamples, filled=False, contour_colors=plt.cm.Set1.colors)
    plt.savefig('testsine_posterior.pdf', bbox_inches='tight')
    plt.close()
    

if __name__ == '__main__':
    parser = argparse.ArgumentParser()

    parser.add_argument('--contrast', type=int, default=100,
                        help="Signal-to-Noise level")
    parser.add_argument('--ndata', type=int, default=40,
                        help="Number of simulated data points")
    parser.add_argument('--sampler', required=True, choices=['goodman-weare', 'slice'])
    args = parser.parse_args()
    main(args)


examples/gauss_example.py
import argparse
import numpy as np
from numpy import pi, sin, log
import matplotlib.pyplot as plt
import logging, sys

def main(args):

    np.random.seed(2)
    ndim = args.x_dim
    sigma = np.logspace(-1, -10, ndim)
    means = np.ones(ndim) * 0.5
    paramnames = ['x%d' % (i+1) for i in range(ndim)]
    ndim = len(paramnames)
    
    def loglike(params):
        logl = (-0.5 * log(2 * pi * sigma.reshape((1, -1))**2) - 0.5 * ((params - means.reshape((1, -1))) / sigma.reshape((1, -1)))**2).sum(axis=1)
        return logl
    
    from autoemcee import ReactiveAffineInvariantSampler, create_logger
    
    create_logger('autoemcee', level=logging.DEBUG)
    sampler = ReactiveAffineInvariantSampler(paramnames, loglike, vectorized=True)
    
    sampler.run(
        #num_global_samples=10000,
        #num_chains=4,
        #num_walkers=10,
        #max_improvement_loops=40,
    )
    if sampler.mpi_rank != 0:
        return
    sampler.print_results()
    
    from getdist import MCSamples, plots

    samples_g = MCSamples(samples=sampler.results['samples'],
                           names=sampler.results['paramnames'],
                           label='Gaussian',
                           settings=dict(smooth_scale_2D=1))

    mcsamples = [samples_g]
    g = plots.get_subplot_plotter(width_inch=8)
    g.settings.num_plot_contours = 3
    g.triangle_plot(mcsamples, filled=False, contour_colors=plt.cm.Set1.colors)
    plt.savefig('testgauss_posterior.pdf', bbox_inches='tight')
    plt.close()


if __name__ == '__main__':
    parser = argparse.ArgumentParser()

    parser.add_argument('--x_dim', type=int, default=2,
                        help="Dimensionality")
    parser.add_argument('--sampler', required=True, choices=['goodman-weare', 'slice'])

    args = parser.parse_args()
    main(args)


examples/eggbox_example.py
import argparse
import numpy as np
from numpy import pi, cos, log
import matplotlib.pyplot as plt
import logging, sys

def main(args):

    np.random.seed(2)
    ndim = 2
    paramnames = ['x%d' % (i+1) for i in range(ndim)]
    ndim = len(paramnames)
    
    def loglike(z):
        chi = (cos(z / 2.)).prod(axis=1)
        return (2. + chi)**5

    def transform(x):
        return x * 10 * pi
    
    from autoemcee import ReactiveAffineInvariantSampler, create_logger
    
    create_logger('autoemcee', level=logging.DEBUG)
    sampler = ReactiveAffineInvariantSampler(
        paramnames, loglike, transform=transform, 
        vectorized=True,
        sampler=args.sampler)
    
    sampler.run(
        #num_chains=4,
        #max_improvement_loops=2,
        max_ncalls=1000000,
        min_autocorr_times=1,
    )
    if sampler.mpi_rank != 0:
        return
    sampler.print_results()
    
    from getdist import MCSamples, plots

    samples_g = MCSamples(samples=sampler.results['samples'],
                           names=sampler.results['paramnames'],
                           label='Gaussian')
                           #settings=dict(smooth_scale_2D=1))

    mcsamples = [samples_g]
    g = plots.get_subplot_plotter(width_inch=8)
    g.settings.num_plot_contours = 3
    g.triangle_plot(mcsamples, filled=False, contour_colors=plt.cm.Set1.colors)
    plt.savefig('testeggbox_posterior.pdf', bbox_inches='tight')
    plt.close()
    

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--sampler', required=True, choices=['goodman-weare', 'slice'])
    args = parser.parse_args()
    main(args)


autoemcee.py
"""Calculates the Bayesian evidence and posterior samples of arbitrary monomodal models."""

from __future__ import print_function
from __future__ import division

import os
import sys
import logging
import warnings
import corner

import numpy as np

import emcee
import arviz

__all__ = ['ReactiveAffineInvariantSampler']
__author__ = """Johannes Buchner"""
__email__ = 'johannes.buchner.acad@gmx.com'
__version__ = '0.1.1'


# Some parts are from the nnest library by Adam Moss (https://github.com/adammoss/nnest)
def create_logger(module_name, log_dir=None, level=logging.INFO):
    """
    Set up the logging channel `module_name`.

    Append to ``debug.log`` in `log_dir` (if not ``None``).
    Write to stdout with output level `level`.

    If logging handlers are already registered, no new handlers are
    registered.
    """
    logger = logging.getLogger(str(module_name))
    logger.setLevel(logging.DEBUG)
    first_logger = logger.handlers == []
    if log_dir is not None and first_logger:
        # create file handler which logs even debug messages
        handler = logging.FileHandler(os.path.join(log_dir, 'debug.log'))
        formatter = logging.Formatter(
            '%(asctime)s [{}] [%(levelname)s] %(message)s'.format(module_name),
            datefmt='%H:%M:%S')
        handler.setFormatter(formatter)
        handler.setLevel(logging.DEBUG)
        logger.addHandler(handler)
    if first_logger:
        # if it is new, register to write to stdout
        handler = logging.StreamHandler(sys.stdout)
        handler.setLevel(level)
        formatter = logging.Formatter('[{}] %(message)s'.format(module_name))
        handler.setFormatter(formatter)
        #logger.setLevel(level)
        logger.addHandler(handler)

    return logger


def vectorize(function):
    """Vectorize likelihood or prior_transform function."""
    def vectorized(args):
        """ vectorized version of function"""
        return np.asarray([function(arg) for arg in args])

    vectorized.__name__ = function.__name__
    return vectorized


class ReactiveAffineInvariantSampler(object):
    """Emcee sampler with reactive exploration strategy."""

    def __init__(self,
                 param_names,
                 loglike,
                 transform=None,
                 num_test_samples=2,
                 vectorized=False,
                 sampler='goodman-weare',
                 ):
        """Initialise sampler.

        Parameters
        -----------
        param_names: list of str, names of the parameters.
            Length gives dimensionality of the sampling problem.

        loglike: function
            log-likelihood function.
            Receives multiple parameter vectors, returns vector of likelihood.
        transform: function
            parameter transform from unit cube to physical parameters.
            Receives multiple cube vectors, returns multiple parameter vectors.
        sampler: str
            if 'goodman-weare': use Goodman & Weare's affine invariant MCMC ensemble sampler
            if 'slice': use Karamanis & Beutler (2020)'s ensemble slice sampler
        vectorized: bool
            if true, likelihood and transform receive arrays of points, and return arrays

        num_test_samples: int
            test transform and likelihood with this number of
            random points for errors first. Useful to catch bugs.
        """
        self.paramnames = param_names
        x_dim = len(self.paramnames)

        self.sampler = 'reactive-importance'
        self.x_dim = x_dim

        self.ncall = 0
        self.use_mpi = False
        self.sampler = sampler
        if sampler not in ('goodman-weare', 'slice'):
            raise ValueError("sampler needs to be one of ('goodman-weare', 'slice')")
        try:
            from mpi4py import MPI
            self.comm = MPI.COMM_WORLD
            self.mpi_size = self.comm.Get_size()
            self.mpi_rank = self.comm.Get_rank()
            if self.mpi_size > 1:
                self.use_mpi = True
                self._setup_distributed_seeds()
        except Exception:
            self.mpi_size = 1
            self.mpi_rank = 0
        
        self.log = self.mpi_rank == 0
        self.logger = create_logger('autoemcee')

        if not vectorized:
            loglike = vectorize(loglike)
            if transform is not None:
                transform = vectorize(transform)

        self.ncall = 0
        self._set_likelihood_function(transform, loglike, num_test_samples)

    def _set_likelihood_function(self, transform, loglike, num_test_samples, make_safe=False):
        """Store the transform and log-likelihood functions.

        Tests with `num_test_samples` whether they work and give the correct output.

        if make_safe is set, make functions safer by accepting misformed
        return shapes and non-finite likelihood values.
        """
        # do some checks on the likelihood function
        # this makes debugging easier by failing early with meaningful errors
        
        # test with num_test_samples random points
        u = np.random.uniform(size=(num_test_samples, self.x_dim))
        p = transform(u) if transform is not None else u
        assert p.shape == (num_test_samples, self.x_dim,), ("Error in transform function: returned shape is %s, expected %s" % (p.shape, self.x_dim))
        logl = loglike(p)
        assert np.logical_and(u > 0, u < 1).all(), ("Error in transform function: u was modified!")
        assert logl.shape == (num_test_samples,), ("Error in loglikelihood function: returned shape is %s, expected %s" % (logl.shape, num_test_samples))
        assert np.isfinite(logl).all(), ("Error in loglikelihood function: returned non-finite number: %s for input u=%s p=%s" % (logl, u, p))

        self.loglike = loglike

        if transform is None:
            self.transform = lambda x: x
        else:
            self.transform = transform

    def _setup_distributed_seeds(self):
        if not self.use_mpi:
            return
        seed = 0
        if self.mpi_rank == 0:
            seed = np.random.randint(0, 1000000)

        seed = self.comm.bcast(seed, root=0)
        if self.mpi_rank > 0:
            # from http://arxiv.org/abs/1005.4117
            seed = int(abs(((seed * 181) * ((self.mpi_rank - 83) * 359)) % 104729))
            # print('setting seed:', self.mpi_rank, seed)
            np.random.seed(seed)

    def _emcee_logprob(self, u):
        mask = np.logical_and((u > 0).all(axis=1), (u < 1).all(axis=1))
        L = -np.inf * np.ones(len(u))

        p = self.transform(u[mask, :])
        L[mask] = self.loglike(p)
        return L

    def find_starting_walkers(self, num_global_samples, num_walkers):
        assert num_global_samples > num_walkers, (num_global_samples, num_walkers)

        ndim, loglike, transform = self.x_dim, self.loglike, self.transform

        if self.log:
            self.logger.debug("global sampling for starting point ...")

        u = np.random.uniform(size=(num_global_samples, ndim))
        p = transform(u)
        L = loglike(p)
        
        # find indices of the highest likelihood ones
        i = np.argsort(L)[::-1][:num_walkers]
        
        return u[i, :], p[i, :], L[i]


    def run(self,
            num_global_samples=10000,
            num_chains=4,
            num_walkers=None,
            max_ncalls=1000000,
            max_improvement_loops=4,
            num_initial_steps=100,
            min_autocorr_times=0,
            progress=True):
        """Sample until MCMC chains have converged.

        The steps are:

        1. Draw *num_global_samples* from prior. The highest *num_walkers* points are selected.
        2. Set *num_steps* to *num_initial_steps*
        3. Run *num_chains* MCMC ensembles for *num_steps* steps
        4. For each walker chain, compute auto-correlation length (Convergence requires *num_steps*/autocorrelation length > *min_autocorr_times*)
        5. For each parameter, compute geweke convergence diagnostic (Convergence requires \|z\| < 2)
        6. For each ensemble, compute gelman-rubin rank convergence diagnostic (Convergence requires rhat<1.2)
        7. If converged, stop and return results. 
        8. Increase *num_steps* by 10, and repeat from (3) up to *max_improvement_loops* times.


        Parameters
        ----------

        num_global_samples: int
            Number of samples to draw from the prior to
        num_chains: int
            Number of independent ensembles to run. If running with MPI,
            this is set to the number of MPI processes.
        num_walkers: int
            Ensemble size. If None, max(100, 4 * dim) is used
        max_ncalls: int
            Maximum number of likelihood function evaluations
        num_initial_steps: int
            Number of sampler steps to take in first iteration
        max_improvement_loops: int
            Number of times MCMC should be re-attempted (see above)
        min_autocorr_times: float
            if positive, additionally require for convergence that the 
            number of samples is larger than the *min_autocorr_times*
            times the autocorrelation length.
        progress: bool
            if True, show progress bars

        """
        
        if num_walkers is None:
            num_walkers = max(100, 4 * self.x_dim)

        num_steps = num_initial_steps
        if self.use_mpi:
            num_chains = self.mpi_size
            num_chains_here = 1
        else:
            num_chains_here = num_chains

        if self.log:
            self.logger.info("finding starting points and running initial %d MCMC steps" % (num_steps))
        
        self.samplers = []
        for chain in range(num_chains_here):
            u, p, L = self.find_starting_walkers(num_global_samples, num_walkers)
            
            if self.sampler == 'goodman-weare':
                sampler = emcee.EnsembleSampler(num_walkers, self.x_dim, self._emcee_logprob, vectorize=True)
            elif self.sampler == 'slice':
                import zeus
                sampler = zeus.sampler(nwalkers=num_walkers, ndim=self.x_dim, logprob_fn=self._emcee_logprob, vectorize=True)
            self.samplers.append(sampler)
            sampler.run_mcmc(u, num_steps, progress=self.log and progress)
        
        self.ncall += num_chains * (num_global_samples + (num_steps + 1) * num_walkers)
        
        for it in range(max_improvement_loops):
            if self.log:
                self.logger.debug("checking convergence (iteration %d) ..." % (it+1))
            converged = True
            # check state of chains:
            for sampler in self.samplers:
                chain = sampler.get_chain()
                assert chain.shape == (num_steps, num_walkers, self.x_dim), (chain.shape, (num_steps, num_walkers, self.x_dim))
                accepts = (chain[1:, :, :] != chain[:-1, :, :]).any(axis=2).sum(axis=0)
                assert accepts.shape == (num_walkers,)
                if self.log:
                    i = np.argsort(accepts)
                    self.logger.debug("acceptance rates: %s (worst few)", accepts[i[:8]] * 100. / num_steps)
                
                # flatchain = sampler.get_chain(flat=True)
                
                # diagnose this chain
                
                # 0. analyse each variable
                max_autocorrlength = 1
                for i in range(self.x_dim):
                    chain_variable = chain[:, :, i]
                    # 1. treat each walker as a independent chain
                    try:
                        for w in range(num_walkers):
                            chain_walker = chain_variable[:, w]
                            with warnings.catch_warnings():
                                warnings.simplefilter("ignore")
                                tau = emcee.autocorr.integrated_time(chain_walker, c=5, tol=50, quiet=False)
                                tau = max(tau, 1)
                            max_autocorrlength = max(max_autocorrlength, tau)
                            if len(chain_walker) / tau < min_autocorr_times:
                                self.logger.debug("autocorrelation is long for parameter '%s': tau=%.1f -> %dx lengths" % (self.paramnames[i], tau, num_steps / tau))
                                converged = False
                                break
                    except emcee.autocorr.AutocorrError:
                        max_autocorrlength = np.inf
                        if min_autocorr_times > 0:
                            self.logger.debug("autocorrelation is too long for parameter '%s' to be estimated" % (self.paramnames[i]))
                            converged = False
                            break

                    if not converged:
                        break

                    # secondly, detect drift with geweke
                    a = sampler.get_chain(flat=True)[:num_steps // 4, i]
                    b = sampler.get_chain(flat=True)[-num_steps // 4:, i]
                    geweke_z = (a.mean() - b.mean()) / (np.var(a) + np.var(b))**0.5
                    if geweke_z > 2.:
                        self.logger.debug("geweke drift (z=%.1f) detected for parameter '%s'" % (geweke_z, self.paramnames[i]))
                        converged = False
                    
                self.logger.debug("autocorrelation length: tau=%.1f -> %dx lengths" % (max_autocorrlength, num_steps / max_autocorrlength))
                if not converged:
                    break
            
            # merge converged across MPI chains
            if self.use_mpi:
                recv_converged = self.comm.gather(converged, root=0)
                converged = all(self.comm.bcast(recv_converged, root=0))
            
            
            if converged:
                # finally, gelman-rubin diagnostic on chains
                chains = np.asarray([sampler.get_chain(flat=True) for sampler in self.samplers])
                if self.use_mpi:
                    recv_chains = self.comm.gather(chains, root=0)
                    chains = np.concatenate(self.comm.bcast(recv_chains, root=0))

                assert chains.shape == (num_chains, num_steps * num_walkers, self.x_dim), (chains.shape, (num_chains, num_steps * num_walkers, self.x_dim))

                rhat = arviz.rhat(arviz.convert_to_dataset(chains)).x.data
                if self.log:
                    self.logger.debug("rhat chain diagnostic: %s (<1.2 is good)", rhat)
                converged = np.all(rhat < 1.2)

                if self.use_mpi:
                    converged = self.comm.bcast(converged, root=0)
            
            if converged:
                if self.log:
                    self.logger.info("converged!!!")
                break
            
            if self.ncall > max_ncalls:
                if self.log:
                    self.logger.warn("maximum number of likelihood calls reached")
                break
                
            
            if self.log:
                self.logger.info("not converged yet at iteration %d" % (it+1))
            #self.logger.error("error at iteration %d" % (it+1))
            last_num_steps = num_steps
            num_steps = int(last_num_steps * 10)
            
            self.ncall += num_chains * (num_steps + 1 + last_num_steps) * num_walkers
            if self.log:
                self.logger.info("Running %d MCMC steps ..." % (num_steps))
            for sampler in self.samplers:
                chain = sampler.get_chain(flat=True)
                # get a scale small compared to the width of the current posterior
                std = np.clip(chain.std(axis=0) * 1e-5, a_min=1e-30, a_max=1)
                assert std.shape == (self.x_dim,), std.shape
                
                # sample num_walkers points from chain, proportional to likelihood
                L = sampler.get_log_prob(flat=True)
                p = np.exp(L - L.max())
                i = np.random.choice(len(L), size=num_walkers, p=p / p.sum())
                # select points
                u = np.clip(chain[i, :], 1e-10, 1 - 1e-10)
                # add a bit of noise
                noise = np.random.normal(0, std, size=(num_walkers, self.x_dim))
                u = u + noise
                # avoid border
                u = np.clip(u, 1e-10, 1 - 1e-10)
                assert u.shape == (num_walkers, self.x_dim), (u.shape, (num_walkers, self.x_dim))
                
                if self.log:
                    self.logger.info("Starting at %s +- %s", u.mean(axis=0), u.std(axis=0))
                sampler.reset()
                #self.logger.info("not converged yet at iteration %d" % (it+1))
                sampler.run_mcmc(u, last_num_steps, progress=self.log)
                last_samples = sampler.get_chain()[-1, :, :]
                assert last_samples.shape == (num_walkers, self.x_dim), (last_samples.shape, (num_walkers, self.x_dim))
                sampler.reset()
                sampler.run_mcmc(last_samples, num_steps, progress=self.log and progress)

        if self.transform is None:
            eqsamples = np.concatenate([sampler.get_chain(flat=True) for sampler in self.samplers])
        else:
            eqsamples = np.concatenate([self.transform(sampler.get_chain(flat=True)) for sampler in self.samplers])

        if self.use_mpi:
            recv_eqsamples = self.comm.gather(eqsamples, root=0)
            eqsamples = np.concatenate(self.comm.bcast(recv_eqsamples, root=0))
        
        self.results = dict(
            paramnames=self.paramnames,
            posterior=dict(
                mean=eqsamples.mean(axis=0).tolist(),
                stdev=eqsamples.std(axis=0).tolist(),
                median=np.percentile(eqsamples, 50, axis=0).tolist(),
                errlo=np.percentile(eqsamples, 15.8655, axis=0).tolist(),
                errup=np.percentile(eqsamples, 84.1345, axis=0).tolist(),
            ),
            samples=eqsamples,
            ncall = self.ncall,
        )
        return self.results
        
    def print_results(self):
        "" "Give summary of marginal likelihood and parameters."" "
        if self.log:
            print()
            for i, p in enumerate(self.paramnames):
                v = self.results['samples'][:, i]
                sigma = v.std()
                med = v.mean()
                if sigma == 0:
                    i = 3
                else:
                    i = max(0, int(-np.floor(np.log10(sigma))) + 1)
                fmt = '%%.%df' % i
                fmts = '\t'.join(['    %-20s' + fmt + " +- " + fmt])
                print(fmts % (p, med, sigma))

    def plot(self, **kwargs):
        if self.log:
            corner.corner(
                self.results['samples'],
                labels=self.results['paramnames'],
                show_titles=True)
